stdlib/public/core/RangeReplaceableCollection.swift.gyb

```diff
@@ -1212,6 +1212,31 @@ public func += <
   lhs.append(contentsOf: rhs)
 }
 
+extension RangeReplaceableCollection {
+  /// Returns a new collection of the same type containing, in order, the
+  /// elements of the original collection that satisfy the given predicate.
+  ///
+  /// In this example, `filter(_:)` is used to include only names shorter than
+  /// five characters.
+  ///
+  ///     let cast = ["Vivien", "Marlon", "Kim", "Karl"]
+  ///     let shortNames = cast.filter { $0.count < 5 }
+  ///     print(shortNames)
+  ///     // Prints "["Kim", "Karl"]"
+  ///
+  /// - Parameter isIncluded: A closure that takes an element of the
+  ///   sequence as its argument and returns a Boolean value indicating
+  ///   whether the element should be included in the returned array.
+  /// - Returns: An array of the elements that `isIncluded` allowed.
+  @_inlineable
+  @available(swift, introduced: 4.0)
+  public func filter(
+    _ isIncluded: (Element) throws -> Bool
+  ) rethrows -> Self {
+    return try Self(self.lazy.filter(isIncluded))
```

Comment:
Here is the snippet I played with:

```swift
@inline(never)
public func iteration<
  C : RangeReplaceableCollection
>(_ xs: C, _ f: (C.Element) -> Bool) -> C {
    var result = C()
    var iterator = xs.makeIterator()
    while let element = iterator.next() {
      if f(element) {
        result.append(element)
      }
    }
    return result
}

@inline(never)
public func lazyFilter<
  C : RangeReplaceableCollection
>(_ xs: C, _ f: (C.Element) -> Bool) -> C {
  return C(xs.lazy.filter(f))
}


for _ in 0..<1_000 {
  /*_ = iteration(Array(0..<10000)) { $0 % 2 == 0 }*/
  _ = lazyFilter(Array(0..<10000)) { $0 % 2 == 0 }
}
```
Commenting one and uncommenting the other gives these results on my machine:

```
# lazyFilter                                                                                                                        
swiftc -swift-version 3 -O ../filter.swift -o ../filter && time ../filter
../filter  0.59s user 0.00s system 99% cpu 0.604 total

# iteration
swiftc -swift-version 3 -O ../filter.swift -o ../filter && time ../filter
../filter  1.45s user 0.00s system 99% cpu 1.462 total

#lazyFilter                                                                                                    
swiftc -swift-version 4 -O ../filter.swift -o ../filter && time ../filter
../filter  0.60s user 0.00s system 99% cpu 0.604 total

# iteration                                                                                                                        
swiftc -swift-version 4 -O ../filter.swift -o ../filter && time ../filter
../filter  1.46s user 0.00s system 99% cpu 1.469 total

```